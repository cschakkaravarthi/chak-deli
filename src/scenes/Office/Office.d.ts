import React, { FC } from 'react';
import { match } from 'react-router';
import { Office as OfficeModel } from '../../shared/models/Office.model';
import Link from '../../shared/models/Link.model';
import PeopleModel from '../../shared/models/People.model';
declare type Props = {
    match?: match;
    links?: Link[];
    facilities?: Link[];
    office?: OfficeModel;
    isLastPage: boolean;
    officeError?: boolean;
    people?: PeopleModel[];
    clearPeople?: () => void;
    facilitiesError?: boolean;
    clearOfficeAction?: () => void;
    clearLinksAction?: () => void;
    fetchFacilities?: (drupalId?: number) => void;
    fetchOffice?: (officeId: string | number) => Promise<void>;
    fetchPeople?: <T>(sitecode: string, startsWith?: string, limit?: number) => T;
};
export declare const Office: FC<Props>;
declare const _default: import("react-redux").ConnectedComponent<React.FC<Props>, Pick<Props, never> & Props>;
export default _default;
//# sourceMappingURL=Office.d.ts.map