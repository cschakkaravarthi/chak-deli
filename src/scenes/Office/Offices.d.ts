import React, { FC } from 'react';
import OfficesPeople, { OfficePeopleCategory } from '../../shared/models/OfficesPeople.model';
import Facet from '../../shared/models/Facet.model';
declare type Props = {
    facets?: Facet;
    selectedFacet?: string;
    fetchOfficesPeople?: <T>() => T;
    officesPeopleList?: OfficesPeople[];
    setFilterFacet: (a: string) => void;
    officeCategories?: OfficePeopleCategory[];
    filterOfficesPeople?: <T>(value: string) => T;
};
export declare const OfficesPeople: FC<Props>;
declare const _default: import("react-redux").ConnectedComponent<React.FC<Props>, Pick<Props, never> & Props>;
export default _default;
//# sourceMappingURL=Offices.d.ts.map