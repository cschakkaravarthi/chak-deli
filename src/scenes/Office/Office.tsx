import React, { FC, useRef, useState, useEffect, PropsWithChildren } from 'react';
import { AppState } from '../../shared/types/genericTypes';
import { match } from 'react-router';
import { Office as OfficeModel } from '../../shared/models/Office.model';
import get from 'lodash.get';
import { fetchOffice, clearOfficeAction } from '../../shared/actions/officeActions';
import { Card, ContentWrapper } from 'umgc_ui_library/lib';
import Container from 'react-bootstrap/Container';
import createDOMPurify from 'dompurify';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Link from '../../shared/models/Link.model';
import { fetchFacilities, clearLinksAction } from '../../shared/actions/linkActions';
import OnError from '../../components/error/Error';
import Facilities from '../../components/facilities/Facilities';
import { connect } from 'react-redux';
import PeopleContainer from '../../components/peopleContainer/PeopleContainer';
import PeopleModel from '../../shared/models/People.model';
import { clearPeople, fetchPeople } from '../../shared/actions/peopleActions';
import { PEOPLE_LIMIT } from '../../constants/constants';
import Error404 from '../Errors/Error404';

import { fetchContent, fetchContentType } from '../../shared/actions/contentActions';
import { toQueryString } from '../../utils/content';
import { ContentQueryParamsType, VariousContentModel } from '../../shared/types/contentTypes';

type Props = {
  match?: match;
  links?: Link[];
  facilities?: Link[];
  office?: OfficeModel;
  isLastPage: boolean;
  officeError?: boolean;
  people?: PeopleModel[];
  clearPeople?: () => void;
  facilitiesError?: boolean;
  clearLinksAction?: () => void;
  clearOfficeAction?: () => void;
  content?: VariousContentModel[];
  fetchContent?: fetchContentType;
  paramSet?: ContentQueryParamsType;
  fetchFacilities?: (drupalId?: number) => void;
  fetchOffice?: (officeId: string | number) => Promise<void>;
  fetchPeople?: <T>(sitecode: string, startsWith?: string, limit?: number) => T;
};

const mapState = (state: AppState, props: Props): Props => {
  const office = state.officeReducers.office;

  const paramSet: ContentQueryParamsType = {
    type: 'page,link,box',
    category: office?.drupal_id
  };

  const content = state.contentReducers.content[toQueryString(paramSet)];

  return {
    ...props,
    office,
    paramSet,
    people: state.peopleReducers.people,
    content: content ? content.content : [],
    isLastPage: state.peopleReducers.isLastPage,
    officeError: state.officeReducers.officeError,
    facilitiesError: state.linkReducers.facilitiesError
  };
};

const actionCreators = {
  fetchOffice,
  clearPeople,
  fetchPeople,
  fetchContent,
  fetchFacilities,
  clearLinksAction,
  clearOfficeAction
};

export const Office: FC<Props> = props => {
  const {
    match,
    office,
    people,
    content,
    paramSet,
    isLastPage,
    fetchPeople,
    officeError,
    fetchOffice,
    clearPeople,
    fetchContent,
    facilitiesError,
    clearLinksAction,
    clearOfficeAction
  } = props;

  const [isFetching, setIsFetching] = useState<boolean>(true);

  const val = useRef<PropsWithChildren<any> | any>();

  useEffect(() => {
    val.current = props;
  }, [props]);

  const urlOfficeId = parseInt(get(match, 'params.id'), 10);

  if (!urlOfficeId) {
    if (!isFetching) {
      return <Error404 />;
    } else {
      return null;
    }
  }

  const [isActive, setIsActive] = useState('');

  let getAdKey: string;

  useEffect(() => {
    getAdKey = '';
    if (urlOfficeId && fetchOffice) {
      fetchOffice(urlOfficeId)
        .then(() => {
          setIsFetching(false);
        })
        .catch(() => {
          setIsFetching(false);
        });
    } else {
      setIsFetching(false);
    }
  }, []);

  const asStringType = [] as string[];

  getAdKey =
    get(office, 'ad_key', asStringType).length > 1
      ? get(office, 'ad_key', asStringType).join()
      : get(office, 'ad_key', asStringType)[0];

  useEffect(() => {
    // this fetches the facilities for now
    fetchContent!(paramSet || {}, undefined, content ? content.length : 0, undefined);
    if (getAdKey && fetchPeople) {
      fetchPeople(getAdKey, undefined, PEOPLE_LIMIT);
    }
  }, [office?.drupal_id]);

  useEffect(() => {
    if (fetchPeople && clearPeople && getAdKey) {
      clearPeople();
      if (isActive) {
        fetchPeople(getAdKey, isActive, PEOPLE_LIMIT);
      } else {
        fetchPeople(getAdKey, undefined, PEOPLE_LIMIT);
      }
    }
  }, [isActive]);

  useEffect(() => {
    return () => {
      if (office) {
        clearPeople!();
        clearLinksAction!();
        clearOfficeAction!();
      }
    };
  }, [val.current?.office?.drupal_id]);

  const handleActiveItem = (letter: string): void => {
    setIsActive(letter);
  };

  const triggerFetchPeople = (): Promise<any> | undefined => {
    if (!isLastPage) {
      return fetchPeople && fetchPeople(getAdKey, isActive, PEOPLE_LIMIT);
    }

    return undefined;
  };

  if (!office) {
    if (!isFetching) {
      return <Error404 />;
    } else {
      return null;
    }
  }

  if (officeError) {
    return <OnError />;
  }

  const clearFilter = (): void => setIsActive('');

  const facilities = content;

  return (
    <Container className="office my-5">
      <p className="font-size-xl mb-5">
        <b>{office.title}</b>
      </p>
      <Row>
        <Col lg="8">
          <Card
            type="map"
            title="Address"
            titleClass="font-size-sm"
            addressOnMap={get(office, 'addressMap.addressForMap', '')}
            address={
              <div
                className="address-top"
                dangerouslySetInnerHTML={{
                  __html: createDOMPurify.sanitize(get(office, 'addressMap.addressTop', ''))
                }}
              />
            }
          />
        </Col>
        <Col lg="4" className="mt-4 mt-lg-0">
          <ContentWrapper
            wrapperClass="h-100"
            title="Office hours"
            topBorder="darkTurquoise"
            titleClass="font-size-base font-weight-bold"
          >
            <div
              className="font-size-sm"
              dangerouslySetInnerHTML={{
                __html: createDOMPurify.sanitize(get(office, 'office_hours', ''))
              }}
            />
          </ContentWrapper>
        </Col>
      </Row>
      {facilities && !!facilities.length && (
        <>
          <p className="font-size-xl mb-3 mt-5">
            <b>Facilities</b>
          </p>
          <Row>
            {facilities &&
              facilities.map((item: VariousContentModel) => (
                <Col md="6" lg="4" className="my-3" key={item.drupal_id + '_facilitiesCol'}>
                  <Facilities facility={item} facilitiesError={facilitiesError} />
                </Col>
              ))}
          </Row>
        </>
      )}
      {!!people && (
        <PeopleContainer
          office={office}
          people={people}
          isActive={isActive}
          isLastPage={isLastPage}
          clearFilter={clearFilter}
          handleActiveItem={handleActiveItem}
          triggerFetchPeople={triggerFetchPeople}
        />
      )}
    </Container>
  );
};

export default connect(mapState, actionCreators)(Office);
