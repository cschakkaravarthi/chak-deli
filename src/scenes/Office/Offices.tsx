import React, { FC, useEffect } from 'react';
import { connect } from 'react-redux';
import { useHistory } from 'react-router-dom';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Form from 'react-bootstrap/Form';
import Container from 'react-bootstrap/Container';
import { Card } from 'umgc_ui_library/lib/cards/card';
import OfficesPeople, {
  OfficePeopleCategory
} from '../../shared/models/OfficesPeople.model';
import { AppState } from '../../shared/types/genericTypes';
import {
  fetchOfficesPeople,
  filterOfficesPeople
} from '../../shared/actions/officesPeopleActions';
import Facet, { FacetProperty } from '../../shared/models/Facet.model';

type Props = {
  facets?: Facet;
  selectedFacet?: string;
  fetchOfficesPeople?: <T>() => T;
  officesPeopleList?: OfficesPeople[];
  setFilterFacet: (a: string) => void;
  officeCategories?: OfficePeopleCategory[];
  filterOfficesPeople?: <T>(value: string) => T;
};

const mapState = (state: AppState, props: Props): Props => ({
  ...props,
  officesPeopleList: state.officesPeopleReducers.officesPeopleFilteredList,
  officeCategories: state.officesPeopleReducers.officeCategories
});

const actionCreators = {
  fetchOfficesPeople,
  filterOfficesPeople
};

export const OfficesPeople: FC<Props> = props => {
  const {
    officesPeopleList,
    officeCategories,
    fetchOfficesPeople,
    filterOfficesPeople,
    selectedFacet
  } = props;

  useEffect(() => {
    fetchOfficesPeople && fetchOfficesPeople();
  }, []);

  const historyData = useHistory();

  const handleOfficeCardClick = (item: OfficesPeople): void => {
    return historyData.push(`/office/${item && item.drupal_id}`);
  };

  function handleDropDownChange (value: string): void {
    filterOfficesPeople && filterOfficesPeople(value);
  }

  return (
    <Container className="article-list mt-5">
      <Col className="mb-2 article-list-header">
        <Row>
          <p className="font-size-xl">
            <b>Offices</b>
          </p>
        </Row>
        <Row className="article-filter-section my-auto">
          <Col md={6} className="d-flex justify-content-start">
            <Row>
              <p className="font-size-md g600-text-clr text-left">
                <b>Select an office to view directory of people</b>
              </p>
            </Row>
          </Col>
          <Col md={6}>
            <Row className="d-flex justify-content-md-end">
              <Form className="d-flex">
                <Form.Label className="d-flex font-size-sm mr-3 mb-0 align-items-center">
                  Region:
                </Form.Label>
                <Form.Control
                  className="w-auto"
                  size="sm"
                  as="select"
                  id="filter"
                  name="filter"
                  onChange={(e: React.SyntheticEvent) =>
                    handleDropDownChange((e.target as HTMLInputElement).value)
                  }
                  value={selectedFacet}
                >
                  <option value={0}>All</option>
                  {officeCategories &&
                    officeCategories.map((category: FacetProperty) => (
                      <option
                        key={category.drupal_id}
                        value={category.drupal_id}
                      >
                        {category.title}
                      </option>
                    ))}
                </Form.Control>
              </Form>
            </Row>
          </Col>
        </Row>
      </Col>
      <Row>
        {officesPeopleList && (
          <div className="ml-2 p-2 w-100 mb-5 office-people-list">
            <Card
              title=""
              variant="light"
              type="officesListContainer"
              officeData={officesPeopleList}
              onCardClick={(item: OfficesPeople) => handleOfficeCardClick(item)}
            />
          </div>
        )}
      </Row>
    </Container>
  );
};

export default connect(mapState, actionCreators)(OfficesPeople);
