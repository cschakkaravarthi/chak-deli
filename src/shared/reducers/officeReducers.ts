import { Reducer } from 'redux';

import {
  LOAD_OFFICE_BY_ID,
  LOAD_OFFICE_BY_ID_ERROR,
  CLEAR_OFFICE_REDUCER,
  OfficeState
} from '../types/officeTypes';

export const initialState = {
  office: undefined,
  officeError: false
};

const officeReducers: Reducer<OfficeState, any> = (
  state = initialState,
  action
) => {
  switch (action.type) {
    case LOAD_OFFICE_BY_ID:
      return {
        ...state,
        office: action.office,
        officeError: false
      };
    case LOAD_OFFICE_BY_ID_ERROR:
      return {
        ...state,
        officeError: true
      };
    case CLEAR_OFFICE_REDUCER:
      return {
        ...state,
        office: initialState.office
      };
    default:
      return state;
  }
};

export default officeReducers;
