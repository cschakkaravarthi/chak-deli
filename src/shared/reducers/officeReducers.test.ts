import officeReducers, { initialState } from './officeReducers';

import {
  LOAD_OFFICE_BY_ID,
  LOAD_OFFICE_BY_ID_ERROR
} from '../types/officeTypes';

import { dummyOffice } from '../models/Office.model';

describe('officeReducers', () => {
  it('Handles default case', () => {
    const state = officeReducers(undefined, {});
    expect(state).toEqual(initialState);
  });

  it('Handles LoadOfficeAction action', () => {
    const action = { type: LOAD_OFFICE_BY_ID, office: dummyOffice };
    const state = officeReducers(initialState, action);
    expect(state).toMatchSnapshot();
  });

  it('Handles OfficeError action', () => {
    const action = { type: LOAD_OFFICE_BY_ID_ERROR, profileDataError: true };
    const state = officeReducers(initialState, action);
    expect(state).toMatchSnapshot();
  });
});
