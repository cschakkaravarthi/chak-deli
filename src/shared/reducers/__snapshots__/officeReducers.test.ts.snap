// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`officeReducers Handles LoadOfficeAction action 1`] = `
Object {
  "office": Object {
    "ad_key": Array [
      "USWHB",
    ],
    "addressMap": Object {
      "addressForMap": "21301 Burbank Blvd Woodland Hills, CA 91367",
      "addressTop": "<p>21301 Burbank Blvd</p>

<p>Woodland Hills, CA 91367</p>
",
    },
    "categories": Array [
      Object {
        "drupal_id": 118,
        "icon": null,
        "title": "North America",
        "weight": 0,
      },
    ],
    "drupal_id": 135,
    "language": "en",
    "location": Array [
      Object {
        "drupal_id": 13,
        "icon": null,
        "title": "Los Angeles Area",
        "weight": 0,
      },
    ],
    "office_hours": "<p>MON - FRI: 7:00AM - 9:00PM</p>

<p>SAT:&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; 8:00AM - 8:00PM</p>

<p>SUN:&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; 8:00AM - 6:00PM</p>
",
    "title": "Woodland Hills",
    "type": "office",
  },
  "officeError": false,
}
`;

exports[`officeReducers Handles OfficeError action 1`] = `
Object {
  "office": undefined,
  "officeError": true,
}
`;
